[mlir][gpu] Disjoint patterns for lowering clustered subgroup reduce

andfau-amd opened the pull request:
Making the existing populateGpuLowerSubgroupReduceToShufflePatterns() function also cover the new "clustered" subgroup reductions is proving to be inconvenient, because certain backends may have more specific lowerings that only cover the non-clustered type, and this creates pass ordering constraints. This commit removes coverage of clustered reductions from this function in favour of a new separate function, which makes controlling the lowering much more straightforward.

kuhar commented on mlir/lib/Dialect/GPU/Transforms/SubgroupReduceLowering.cpp:
```diff
@@ -262,19 +273,31 @@ struct ScalarSubgroupReduceToShuffles final
 private:
   unsigned subgroupSize = 0;
   unsigned shuffleBitwidth = 0;
+  bool matchClustered;
```
nit: set a default value so that it's always initialized

kuhar commented on mlir/lib/Dialect/GPU/Transforms/SubgroupReduceLowering.cpp:
```diff
@@ -262,19 +273,31 @@ struct ScalarSubgroupReduceToShuffles final
   LogicalResult matchAndRewrite(gpu::SubgroupReduceOp op,
                                 PatternRewriter &rewriter) const override {
+    if (op.getClusterSize().has_value() != matchClustered) {
+      if (matchClustered)
+        return rewriter.notifyMatchFailure(
+            op, "op is non-clustered but pattern is configured to only match "
+                "clustered ops");
+      else
+        return rewriter.notifyMatchFailure(
+            op, "op is clustered but pattern is configured to only match "
+                "non-clustered ops");
```
```suggestion
      if (matchClustered) {
        return rewriter.notifyMatchFailure(
            op, "op is non-clustered but pattern is configured to only match "
                "clustered ops");
      } else {
        return rewriter.notifyMatchFailure(
            op, "op is clustered but pattern is configured to only match "
                "non-clustered ops");
     }
```

kuhar commented on mlir/lib/Dialect/GPU/Transforms/SubgroupReduceLowering.cpp:
```diff
@@ -343,6 +366,7 @@ struct VectorSubgroupReduceToShuffles final
 private:
   unsigned subgroupSize = 0;
   unsigned shuffleBitwidth = 0;
+  bool matchClustered;
```
Same here

kuhar commented on mlir/lib/Dialect/GPU/Transforms/SubgroupReduceLowering.cpp:
```diff
@@ -210,13 +210,24 @@ Value createSubgroupShuffleReduction(OpBuilder &builder, Location loc,
 
   LogicalResult matchAndRewrite(gpu::SubgroupReduceOp op,
                                 PatternRewriter &rewriter) const override {
+    if (op.getClusterSize().has_value() != matchClustered) {
+      if (matchClustered)
+        return rewriter.notifyMatchFailure(
+            op, "op is non-clustered but pattern is configured to only match "
+                "clustered ops");
+      else
+        return rewriter.notifyMatchFailure(
+            op, "op is clustered but pattern is configured to only match "
+                "non-clustered ops");
+    }
```
also here

kuhar commented on mlir/lib/Dialect/GPU/Transforms/SubgroupReduceLowering.cpp:
```diff
@@ -210,13 +210,25 @@ Value createSubgroupShuffleReduction(OpBuilder &builder, Location loc,
+        return rewriter.notifyMatchFailure(
+            op, "op is non-clustered but pattern is configured to only match "
+                "clustered ops");
+      } else {
+        return rewriter.notifyMatchFailure(
+            op, "op is clustered but pattern is configured to only match "
+                "non-clustered ops");
+      }
```
Ah, now I see that we don't need the `else`
```suggestion
      }
      return rewriter.notifyMatchFailure(
          op, "op is clustered but pattern is configured to only match "
              "non-clustered ops");
```

andfau-amd replied:
I don't really like the asymmetry in this version. I've pushed another approach, does it look better?

kuhar commented on mlir/lib/Dialect/GPU/Transforms/SubgroupReduceLowering.cpp:
```diff
@@ -262,19 +274,32 @@ struct ScalarSubgroupReduceToShuffles final
+        return rewriter.notifyMatchFailure(
+            op, "op is non-clustered but pattern is configured to only match "
+                "clustered ops");
+      } else {
+        return rewriter.notifyMatchFailure(
+            op, "op is clustered but pattern is configured to only match "
+                "non-clustered ops");
+      }
```
```suggestion
      }
      return rewriter.notifyMatchFailure(
          op, "op is clustered but pattern is configured to only match "
              "non-clustered ops");
```

andfau-amd committed:
[mlir][gpu] Disjoint patterns for lowering clustered subgroup reduce

Making the existing populateGpuLowerSubgroupReduceToShufflePatterns()
function also cover the new "clustered" subgroup reductions is proving
to be inconvenient, because certain backends may have more specific
lowerings that only cover the non-clustered type, and this creates pass
ordering constraints. This commit removes coverage of clustered
reductions from this function in favour of a new separate function,
which makes controlling the lowering much more straightforward.
